src/gwt/src/org/rstudio/studio/client/common/mathjax/MathJax.java

```diff
@@ -112,6 +100,41 @@ public void onLoaded(boolean alreadyLoaded)
             });
          }
       }));
+      
+      handlers_.add(docDisplay_.addCursorChangedHandler(new CursorChangedHandler()
+      {
+         @Override
+         public void onCursorChanged(CursorChangedEvent event)
+         {
+            Position position = event.getPosition();
+            
+            // re-render line widgets eagerly when editing within latex chunk
+            Range range = MathJaxUtil.getLatexRange(docDisplay_);
+            if (range != null && range.contains(position) && MathJaxUtil.isSelectionWithinLatexChunk(docDisplay_))
+            {
+               renderTimer_.schedule(range, true, 50);
+               return;
+            }
+            
+            // end a render session for inline latex
+            if (anchor_ == null || !anchor_.getRange().contains(position))
+            {
+               endRender();
+               return;
+            }
+            
+            // re-render inline latex (deferred so that anchor can update)
+            Scheduler.get().scheduleDeferred(new ScheduledCommand()
```

Comment:
Consider hooking up a single shot render finished listener -- IIRC you can use those to know deterministically when the document state is consistent